### ISSUE TITLE
Obsidian Registration Entry: Object reference not set to an instance of an object.

### ISSUE THREAD

**Issue by JoshuaMaiorino** [SparkDevNetwork/Rock#5366]
### Please go through all the tasks below

- [X] Check this box only after you have successfully completed both the above tasks

### Please provide a brief description of the problem. Please do not forget to attach the relevant screenshots from your side.

When certain people are attempting to register their children for event, when clicking the 'Finish' Button, an Object Reference error is being thrown.  This issue is happening intermittently.

### Expected Behavior

Either the registration should go through, or a more detailed error message explaining why it didn't should be shown.

### Actual Behavior

No Registration gets saved to the DB, and we're just getting the Object Reference error.

### Steps to Reproduce

Go to the Registration and attempt to register children for the registration.

here's a sanitized version of the payload being submitted.
```{
    "__context": {
        "pageParameters": {
            "RegistrationInstanceId": "2021"
        }
    },
    "args": {
        "registrationSessionGuid": "533e4fef-a0cc-4553-963a-e0a0f9bb18e6",
        "gatewayToken": "",
        "savedAccountGuid": null,
        "discountCode": "",
        "fieldValues": {},
        "registrar": {
            "nickName": "***",
            "lastName": "********",
            "email": "***.*******@gmail.com",
            "updateEmail": true,
            "familyGuid": "533e4fef-a0cc-4553-963a-e0a0f9bb18e6"
        },
        "registrants": [
            {
                "isOnWaitList": false,
                "familyGuid": "533e4fef-a0cc-4553-963a-e0a0f9bb18e6",
                "personGuid": "533e4fef-a0cc-4553-963a-e0a0f9bb18e6",
                "fieldValues": {
                    "5cf3be2f-1940-4411-8ac9-5e10828ea55b": "***",
                    "0eaf2d92-0df7-4f6d-9bd9-9a667135005a": "******",
                    "cbf92f06-4f81-4195-be1f-ed971fc8bfab": {
                        "street1": "*********",
                        "street2": "",
                        "city": "********",
                        "state": "**",
                        "postalCode": "******",
                        "country": null
                    },
                    "44245161-a789-4859-8417-c1f3cf3595ae": "(***) ***-****",
                    "ef94f1a0-c0f6-4d9b-9e0b-57074627b715": {
                        "day": 1,
                        "month": 1,
                        "year": 1900
                    },
                    "ca0e684b-46d9-40a7-9282-15c373c0c259": "*****************",
                    "fd783f44-a4b7-4e48-ad82-26482a5e4fea": "S",
                    "e611f5df-5771-4fb7-b6ff-0c20f4fe9fa0": "9th",
                    "4ccea4aa-598a-425a-9004-4ad7fc6ce993": "************",
                    "4729140c-30fd-4ab3-9659-95a6fd23e4dc": "(***) ***-****",
                    "a069012a-d527-4c18-8823-7516e402a89a": "************************",
                    "c37e5435-acc0-493d-8712-fbf6d6ac3d6c": "********"
                },
                "feeItemQuantities": {
                    "9bbee7db-9f4d-4260-a8a1-b00702639a71": 1,
                    "675cea1b-f542-4c3a-b5e9-942c885c787a": 0,
                    "5b5a78f7-43e6-4328-bf55-a9523991e6dd": 0,
                    "54ad9150-a21d-4ccf-aac3-2c3f209d554d": 0,
                    "b5a44b40-0653-49e8-b39e-0db97f879263": 0,
                    "a5fe2792-b61c-449b-ad59-ac05b2ed05f8": 0,
                    "13667aa3-25da-4f1c-8079-606c2d6a7f6d": 0,
                    "27fa6546-4347-4207-9186-8cf78de7e1ed": 0,
                    "68686c5a-c053-4349-af22-0ff485d463da": 0
                },
                "signatureData": null,
                "guid": "533e4fef-a0cc-4553-963a-e0a0f9bb18e6"
            },
            {
                "isOnWaitList": false,
                "familyGuid": "533e4fef-a0cc-4553-963a-e0a0f9bb18e6",
                "fieldValues": {
                    "5cf3be2f-1940-4411-8ac9-5e10828ea55b": "***",
                    "0eaf2d92-0df7-4f6d-9bd9-9a667135005a": "*********",
                    "cbf92f06-4f81-4195-be1f-ed971fc8bfab": {
                        "street1": "**********",
                        "street2": "",
                        "city": "******",
                        "state": "**",
                        "postalCode": "******-****",
                        "country": null
                    },
                    "ca0e684b-46d9-40a7-9282-15c373c0c259": "***.**********@gmail.com",
                    "ef94f1a0-c0f6-4d9b-9e0b-57074627b715": {
                        "day": 1,
                        "month": 1,
                        "year": 1900
                    },
                    "44245161-a789-4859-8417-c1f3cf3595ae": "(***) ***-****",
                    "fd783f44-a4b7-4e48-ad82-26482a5e4fea": "M",
                    "e611f5df-5771-4fb7-b6ff-0c20f4fe9fa0": "10th",
                    "4ccea4aa-598a-425a-9004-4ad7fc6ce993": "*********",
                    "4729140c-30fd-4ab3-9659-95a6fd23e4dc": "(***) ***-****",
                    "a069012a-d527-4c18-8823-7516e402a89a": "********************************",
                    "c37e5435-acc0-493d-8712-fbf6d6ac3d6c": "Yes"
                },
                "feeItemQuantities": {
                    "9bbee7db-9f4d-4260-a8a1-b00702639a71": 0,
                    "675cea1b-f542-4c3a-b5e9-942c885c787a": 0,
                    "5b5a78f7-43e6-4328-bf55-a9523991e6dd": 0,
                    "54ad9150-a21d-4ccf-aac3-2c3f209d554d": 0,
                    "b5a44b40-0653-49e8-b39e-0db97f879263": 0,
                    "a5fe2792-b61c-449b-ad59-ac05b2ed05f8": 1,
                    "13667aa3-25da-4f1c-8079-606c2d6a7f6d": 0,
                    "27fa6546-4347-4207-9186-8cf78de7e1ed": 0,
                    "68686c5a-c053-4349-af22-0ff485d463da": 0
                },
                "guid": "533e4fef-a0cc-4553-963a-e0a0f9bb18e6",
                "personGuid": "533e4fef-a0cc-4553-963a-e0a0f9bb18e6"
            }
        ],
        "amountToPayNow": 0,
        "registrationGuid": null
    }
}
```

and here's the stack trace information:

```
System.NullReferenceException | Rock.Blocks | Object reference not set to an instance of an object.
at Rock.Blocks.Event.RegistrationEntry.SubmitRegistration(RockContext rockContext, RegistrationContext context, RegistrationEntryBlockArgs args, String& errorMessage)
   at Rock.Blocks.Event.RegistrationEntry.SubmitRegistration(RegistrationEntryBlockArgs args)
```

### Rock Version

14.2

### Client Culture Setting

en-us

**Comment by JoshuaMaiorino** [comment]
Some other possibly relevant data:

- The registration requires login, however, the exception in Rock doesn't have the user tied to it.  I also noticed places in the code that referenced CurrentPerson.PersonAliasId but couldn't find any place where it's checking for null Current Person.  Though, I know Registrations require a registrar.
- The issue is seems to only be affecting one person.  I received report of the person having the error, and then impersonated and got the same error.  However, Logging in as my self, I was able to successfully register my test child. ( My account is Rock Admin though, while this person has special permissions ).
- The Registration does not requirement payment
- The Registration is using 0 dollar fees with limits on usages.

I could provide any other details someone might think is relevant, but I stepped through the code for a couple hours today, and Couldn't find anything jumping out at me.

The only other thing I noticed was that there are places the `Context` object variable has code that's accessing properties, and there's places in the Code that the Context Object can be returned as null.  
```
                var context = GetContext( rockContext, args, out var errorMessage );

                if ( !errorMessage.IsNullOrWhiteSpace() )
                {
                    return ActionBadRequest( errorMessage );
                } 
```
However, I couldn't find a scenario where context would be null, and `errorMessage` wouldn't be set.  So.

**Comment by dataCollegechurch** [comment]
Have you tried to reproduce this on the demo site?

**Comment by ethan-sparkdevnetwork** [comment]
@JoshuaMaiorino Can you get me the full stack trace? Or a login to your Rock instance? You can reach me on Rocketchat (ethan621).

**Comment by ethan-sparkdevnetwork** [comment]
@JoshuaMaiorino Can you run this query with the problem person's ID and let me know the results?
```
-- Put the person ID here.
DECLARE @personId INT = 


DECLARE @familyGroupTypeId INT = (SELECT [Id] FROM [GroupType] WHERE [Guid] = '790E3215-3B10-442B-AF69-616C0DCB998E')

SELECT * FROM [PersonAlias] WHERE [AliasPersonId] = @personId
SELECT * FROM [PersonAlias] WHERE [PersonId] = @personId

select g.[Name], p.[FirstName], p.LastName, p.NickName, p.Email, p.BirthYear, p.BirthMonth, p.BirthDay, p.BirthDate, pn.Number, l.Street1, l.[City], l.[State]
FROM [Person] p
LEFT JOIN [PhoneNumber] pn ON pn.PersonId = p.[Id]
LEFT JOIN [GroupMember] gm ON gm.[PersonId] = p.[Id]
LEFT JOIN [Group] g ON g.[Id] = gm.[GroupId]
LEFT JOIN [GroupLocation] gl ON gl.[GroupId] = g.[Id]
LEFT JOIN [Location] l ON l.[Id] = gl.LocationId
WHERE p.[Id] = @personId
	AND g.GroupTypeId = @familyGroupTypeId
```

**Comment by ethan-sparkdevnetwork** [comment]
@JoshuaMaiorino I am not able to reproduce this issue and have not received any further information. If this is still a problem then create a new issue with the additional info and reference this one.

**Comment by JoshuaMaiorino** [comment]
@ethan-sparkdevnetwork Sorry, I was off last week.  I ran the SQL listed above and got 1 result for the first select, 5 for the 2nd, and 3 for last.  Name on the alias queries returned null, AliasedDateTime was null on all but one record.  ForeignKey, ForeignGuid,ForeignId and LastVisitDateTime was null for all record.  The Location Birthday and Phone number select statement had no null columns.

Also, the full stack is listed above.
System.NullReferenceException | Rock.Blocks | Object reference not set to an instance of an object.
at Rock.Blocks.Event.RegistrationEntry.SubmitRegistration(RockContext rockContext, RegistrationContext context, RegistrationEntryBlockArgs args, String& errorMessage)
   at Rock.Blocks.Event.RegistrationEntry.SubmitRegistration(RegistrationEntryBlockArgs args)

**Comment by zackdutra** [comment]
We also have had this issue, intermittently exactly as described above. It’s nearly impossible to reproduce, unless you find a person that it happens to specifically.

**Comment by nairdo** [comment]
@JoshuaMaiorino (& @zackdutra) We've made several bug fixes in v16.0 (and v15.2) and one of them in particular ([d3b6cb2](https://github.com/SparkDevNetwork/Rock/commit/d3b6cb2c6e9fb55f04a903c3fc11d15821fc0f19)) seems like it might be _related_ to this issue.  More importantly, with newer Rock updates we're also including the .pdb files which should record a line number in the exception's stack trace when an exception occurs.  That will be extremely valuable with tracking this particular problem down.   

Rock v16.0 is still only in Alpha testing (and v15.2 will start alpha testing afterwards) but do either of you know when you would be planning to move from v14.x to v15.2/v16.0?  (In other words, will it be soon after they are released and/or would you be able to try reproducing this issue on an alpha version in your sandbox that has your particular data?)

**Comment by zackdutra** [comment]
@nairdo we're on 15.1 now, so we can do reasonably soon after release

**Comment by thejasonharvey** [comment]
Josh was submitting this issue on our behalf. He was on staff at BEMA but has had a job change this month. We are planning on moving to 15.3 before the end of the year. We are currently running v15 in our dev site and are planning on moving that forward in our dev environment to check things out before we move forward in production. I don't mind trying to reproduce the issue in our dev site. However, as Zack has stated, it is pretty rare. I believe this issue has only happened to us twice as well. We submitted the bug after the second occurrence because of the nature of the error (in that it could only be resolved by staff/admins), and because we were concerned that it would continue to happen.

In Rocket chat, I have seen something similar being asked about by one other person (in case this helps)  - https://chat.rockrms.com/channel/events-registrations/thread/EjMxWkPCabvtPrXyg?jump=ESnw4kRaZH32CRyDd

If we are able to reproduce the issue I will let you know.

**Comment by TheCoreyHall** [comment]
This is currently happening for several users on an Obsidian registration for my current client. They are on 14.3. Here is the full stack trace. I will continue to investigate as well.
![image](https://github.com/SparkDevNetwork/Rock/assets/38293092/c0503f2d-1894-4a9c-a320-bfb91ff479a2)

**Comment by JRichmond27** [comment]
We just came across this same error message on 15.2. The exception log said the error happened on line [913 of Rock.Blocks\Event\RegistrationEntry.cs](https://github.com/SparkDevNetwork/Rock/blob/fcf3f8031a693cc6e50d846abafdee396cf0b467/Rock.Blocks/Event/RegistrationEntry.cs#L913C66-L913C83)

Here's the full stack trace:
```
   at Rock.Blocks.Event.RegistrationEntry.SubmitRegistration(RockContext rockContext, RegistrationContext context, RegistrationEntryBlockArgs args, String& errorMessage) in \Rock.Blocks\Event\RegistrationEntry.cs:line 913
   at Rock.Blocks.Event.RegistrationEntry.SubmitRegistration(RegistrationEntryBlockArgs args) in \Rock.Blocks\Event\RegistrationEntry.cs:line 458
```
That line appears to be comparing email addresses, so I checked the person's profile and noticed that she does not have an email address. I added her address to her profile, and now I'm just waiting to hear back about whether she still gets the error or not.

The only possible wrench in the troubleshooting is that I also happened to update our instance to 15.3 last night.

**Comment by chead4** [comment]
@JRichmond27 Hi Jeff thanks for sharing those additional details! I'm going to reopen this issue.

**Comment by nairdo** [comment]
Thanks again to everyone -- especially Jeff because the line number in the stack trace allowed us to pinpoint the bug and reproduce it.  Since it would only happen if the logged-in _registrar_ had a NULL email address, it was probably pretty rare to occur.
